core/src/main/java/org/wildfly/plugin/core/DeploymentManager.java

```diff
@@ -0,0 +1,203 @@
+     *
+     * @param deployments a set of deployments to redeploy
+     *
+     * @return the result of the deployment
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    DeploymentResult redeploy(Set<Deployment> deployments) throws IOException;
+
+    /**
+     * Undeploys the deployment from the server.
+     *
+     * @param undeployDescription the description for undeploying the content
+     *
+     * @return the result of the deployment
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    DeploymentResult undeploy(UndeployDescription undeployDescription) throws IOException;
+
+    /**
+     * Undeploys the deployment from the server.
+     *
+     * @param undeployDescriptions the descriptions for undeploying the content
+     *
+     * @return the result of the deployment
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    DeploymentResult undeploy(Set<UndeployDescription> undeployDescriptions) throws IOException;
+
+    /**
+     * Returns the available deployments.
+     *
+     * @return the deployments
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    Set<DeploymentDescription> getDeployments() throws IOException;
+
+    /**
+     * Returns all the deployments on the specified server-group. These deployments may also belong to other server
+     * groups.
+     *
+     * @param serverGroup the server group to get the deployments for
+     *
+     * @return the deployments
+     *
+     * @throws IOException           if a failure occurs communicating with the server
+     * @throws IllegalStateException if the running server is not a managed domain
+     */
+    Set<DeploymentDescription> getDeployments(String serverGroup) throws IOException;
+
+    /**
+     * Returns the names of deployed content.
+     *
+     * @return the names of deployed content
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    Set<String> getDeploymentNames() throws IOException;
+
+    /**
+     * Checks if the deployment is on the server.
+     *
+     * @param name the name of the deployment
+     *
+     * @return {@code true} if the deployment exists otherwise {@code false}
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    boolean hasDeployment(String name) throws IOException;
+
+    /**
+     * Checks if the deployment is on the server.
+     *
+     * @param name        the name of the deployment
+     * @param serverGroup the server group to check for the deployment on
+     *
+     * @return {@code true} if the deployment exists otherwise {@code false}
+     */
+    boolean hasDeployment(String name, String serverGroup) throws IOException;
+
+    /**
+     * A factory to create a new deployment manager
+     */
+    class Factory {
```

Comment:
BTW, seeing factory here, have you considered using dependency injection? As you may know, JSR 330 is available in maven. I am not sure what is the situation like in your other target environments, but at least Gradle seems to have it too.